src/main/java/org/apache/sling/api/resource/FallbackValueMap.java

```diff
@@ -0,0 +1,216 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.sling.api.resource;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import org.apache.sling.api.wrappers.ValueMapDecorator;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class FallbackValueMap implements ValueMap {
```

Comment:
I like the idea and have had such use-cases in the past. I'm not entirely sure the name is ideal - to me fallback suggests that there is one fallback. An alternative name could be MergingValueMap - not sure that's better though.

Thinking a little bit further, maybe we shouldn't expose the implementation class directly, but create a ValueMapUtil class, which has static factory methods to instantiate the respective value map implementation. E.g. `ValueMapUtil#merge(ValueMap... valueMaps)` and if you want caching `ValueMapUtil#cache(ValueMap)` (i.e. `cache(merge(vm1, vm2, vm3))` to get a caching merged ValueMap. WDYT?

Reply:
That's interesting :-) what about doing special "collectors" on a stream of resources?

Reply:
-1 though on two things: 

1. using _merge_ word, as there is no information around the ordered aspect of it. 
2. splitting up the cache aspect of things. the implementation is specifically not supporting mutations, and most of the usages are within one request render. Operations like entrySet or keySet or contains might be expensive if made repeatedly. Not sure we want to leave choice of making a suboptimal usage of this? For the mutable scenario

Reply:
for the name, Fallback*s*ValueMap ?

Reply:
having spent my night over it, i think i like your proposal better @jsedding  :-) still not sure about the name though, nor the caching that stays an implementation detail to me
`ValueMaps.asFIFOValueMap(vm1, ..., vmN)` ?

Reply:
i'm general fine with this, also on needs to get used the naming.

the implementation is not so straightforward as it mixes lazy cache filling and prefilling depending on which method is called first, but i can see the point from a performance point of view.

i'm not sure if all have found the link to this PR, it's hidden down in the second part of a sentence in the JIRA ticket. perhaps it would make sense for API extension to put a dedicated mail on the mailing list with a direct link to the PR.

Reply:
For the naming, maybe `ValueMaps.overlayValueMaps(ValueMap... valueMaps)`? We have the resource merger module, which also implements a similar mechanism. So there is precedent for the word "merge" ;) I don't care too much either way. The name `FallbackValueMap` made **me** think of a single fallback, so I mentioned it. That may not be the case for others.

Regarding the separation of caching, I don't think your argument about the immutability of the FallbackValueMap (or whatever the name will be) doesn't hold. The caller of FallbackValueMap has references to all ValueMaps, so these can potentially be modified. In that event your "view" will have a stale cache. I'm not opposed to having a convenience factory method to create a caching merged VM, but the javadocs need to be clear about the limitation **and** I want a non-caching variant as well. I've run into situations where I could not use an API due to the lack of this kind of separation of concerns. You assume that the use-case is during rendering. However, it is equally likely that a service uses the API to do background operations outside of a request.